packages/core/src/hosted-form/hosted-form-vaulting.ts

```diff
@@ -0,0 +1,128 @@
+import { noop, without } from 'lodash';
+
+import { IframeEventListener } from '../common/iframe';
+
+import { InvalidHostedFormConfigError } from './errors';
+import HostedField from './hosted-field';
+import HostedFormOptions from './hosted-form-options';
+import {
+    HostedFormVaultingData,
+    HostedFormVaultingInstrumentFields,
+} from './hosted-form-vaulting-type';
+import {
+    HostedInputEnterEvent,
+    HostedInputEventMap,
+    HostedInputEventType,
+    HostedInputVaultingSucceededEvent,
+} from './iframe-content';
+
+type HostedFormEventCallbacks = Pick<
+    HostedFormOptions,
+    'onBlur' | 'onCardTypeChange' | 'onFocus' | 'onEnter' | 'onValidate'
+>;
+
+export default class HostedFormVaulting {
+    private _bin?: string;
+    private _cardType?: string;
+
+    constructor(
+        private _fields: HostedField[],
+        private _eventListener: IframeEventListener<HostedInputEventMap>,
+        private _eventCallbacks: HostedFormEventCallbacks,
+    ) {
+        const {
+            onBlur = noop,
+            onCardTypeChange = noop,
+            onFocus = noop,
+            onValidate = noop,
+        } = this._eventCallbacks;
+
+        this._eventListener.addListener(HostedInputEventType.Blurred, ({ payload }) =>
+            onBlur(payload),
+        );
+        this._eventListener.addListener(HostedInputEventType.CardTypeChanged, ({ payload }) =>
+            onCardTypeChange(payload),
+        );
+        this._eventListener.addListener(HostedInputEventType.Focused, ({ payload }) =>
+            onFocus(payload),
+        );
+        this._eventListener.addListener(HostedInputEventType.Validated, ({ payload }) =>
+            onValidate(payload),
+        );
+        this._eventListener.addListener(HostedInputEventType.Entered, this._handleEnter);
+
+        this._eventListener.addListener(
+            HostedInputEventType.CardTypeChanged,
+            ({ payload }) => (this._cardType = payload.cardType),
+        );
+        this._eventListener.addListener(
+            HostedInputEventType.BinChanged,
+            ({ payload }) => (this._bin = payload.bin),
+        );
+    }
+
+    getBin(): string | undefined {
+        return this._bin;
+    }
+
+    getCardType(): string | undefined {
+        return this._cardType;
+    }
+
+    async attach(): Promise<void> {
+        this._eventListener.listen();
+
+        const field = this._getFirstField();
+        const otherFields = without(this._fields, field);
+
+        await field.attach();
+        await Promise.all(otherFields.map((otherField) => otherField.attach()));
+    }
+
+    detach(): void {
+        this._eventListener.stopListen();
+
+        this._fields.forEach((field) => {
+            field.detach();
+        });
+    }
+
+    async submit(payload: {
```

Comment:
It looks like the fields don't have a separate class implementation. Instead, they expose a separate `submit` method for posting to the vaulting endpoint. Could the form also take a similar approach so that the design is consistent with the fields (e.g.: `submitStoredCard`)? We could revisit the design in the future if there's a greater need to create entirely separate implementations for vaulting.

Reply:
It makes sense! 👍
Previously I thought to do it in this way, but somehow I though that it is a bad idea to create a checkout store instance just for creation of the hosted form. But after your comment I re-though it and there is [nothing wrong with it ](https://github.com/bigcommerce/checkout-sdk-js/pull/2230/files#diff-6a1d91e54da7e5712a131b154877196ab8f7a45fc918b26d26321671005b607eR14). So I [added it to the HostedForm class](https://github.com/bigcommerce/checkout-sdk-js/pull/2230/files#diff-40e7580dc4c696da6802c1ee6190abb57a12f19ce78664a5430915d0a92e8756R118) to avoid entirely separate implementations for vaulting.